File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing/ConfigOptions.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Collections.Generic;
using System.Linq;
using Microsoft.CodeAnalysis.Diagnostics;

namespace Microsoft.CodeAnalysis.CSharp.Testing
{
    /// <summary>
    /// Allows adding additional global options
    /// </summary>
    internal class ConfigOptions : AnalyzerConfigOptions
    {
        private readonly AnalyzerConfigOptions _workspaceOptions;
        private readonly Dictionary<string, string> _globalOptions;

        public ConfigOptions(AnalyzerConfigOptions workspaceOptions, List<(string option, string val)> globalOptions)
        {
            _workspaceOptions = workspaceOptions;
            _globalOptions = globalOptions.ToDictionary(t => t.option, t => t.val);
        }

        public override bool TryGetValue(string key, out string value)
            => _workspaceOptions.TryGetValue(key, out value!) || _globalOptions.TryGetValue(key, out value!);
    }
}

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.VisualBasic.SourceGenerators.Testing/ConfigOptions.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis.Diagnostics

Friend Class ConfigOptions
    Inherits AnalyzerConfigOptions

    Private ReadOnly _workspaceOptions As AnalyzerConfigOptions
    Private ReadOnly _globalOptions As Dictionary(Of String, String)

    Public Sub New(workspaceOptions As AnalyzerConfigOptions, globalOptions As List(Of (String, String)))
        _workspaceOptions = workspaceOptions
        _globalOptions = globalOptions.ToDictionary(Function(pair) pair.Item1, Function(pair) pair.Item2)
    End Sub

    Public Overrides Function TryGetValue(key As String, ByRef value As String) As Boolean
        Return _workspaceOptions.TryGetValue(key, value) Or _globalOptions.TryGetValue(key, value)
    End Function

End Class
